### PR TITLE
Combat rework

### PR DESCRIPTION
**Pull request by Irbynx**
Purpose:
* Implement the changes to combat outlined previously

Changes:
* Complete overhaul of combat loop
* Turn based system changed to AP based turn system
* Able to see intents of enemies in combat
* Abilities available to any character in combat are limited to maximum of 8 - respective menu added to phone to assign abilities
* Combat AI is a little smarter now, although improvements can be made
* As a result of the complete overhaul of the abilities, racial moves are not available in full yet.
* Spells are not balanced for this system  either.
* HP formula changed to avoid physique playing a major role in it. Might change the calculations again later, however.
* Mana formula changed to overally bring the amount of mana to 100-120 average values.
* A massive overhaul to unarmed weapons 
* Deprecated special tease attacks
* Deprecated percentage based protection - use blocks instead.
* Did not enable ranged/melee specific damage adjustments yet. Might deprecate.
* Added functionality to calculate damages and resistances to DamageType class to avoid it sitting in Attack class as Attack class will be deprecated soon.
* Added blocking system

Graphical Assets:
* Placeholder images can be replaced with better ones for the moves

Tested:
* 0.2.10.5
* While I tried to test the system, an overhaul of this scale can end up with issues.

Discord handle:
* Irbynx

### CODE REVIEW

**Comment by CognitiveMist** [comment]
Not knowing what was "outlined previously", what is the goal of this change?

Will combat be more random? More strategic? Longer? Shorter? Harder? Easier? Streamlined? Tedious?

File: src/com/lilithsthrone/game/combat/Combat.java [hunk 1]
```diff
@@ -177,7 +179,40 @@ public void initialiseCombat(
 			}
 		}
 
-		combatContent = combatStringBuilder.toString();
+		Main.game.getPlayer().resetSelectedMoves();
+		Main.game.getPlayer().resetMoveCooldowns();
+		Main.game.getPlayer().setRemainingAP(Main.game.getPlayer().getMaxAP(), null, null);
+		for(NPC npc : allCombatants)
```
Comment (picobyte) [inline src/com/lilithsthrone/game/combat/Combat.java]: allCombatants is not yet populated?

File: src/com/lilithsthrone/game/combat/Combat.java [hunk 2]
```diff
@@ -177,7 +179,40 @@ public void initialiseCombat(
...
+		{
+			npc.resetSelectedMoves();
+			npc.resetDefaultMoves(); // Resetting in case the save file was too old and NPC has no moves selected for them.
+			npc.resetMoveCooldowns();
+			npc.setRemainingAP(npc.getMaxAP(), null, null);
+			// Sets up NPC ally/enemy lists that include player
+			List<GameCharacter> npcAllies;
+			List<GameCharacter> npcEnemies;
+			if(allies.contains(npc))
```
Comment (picobyte) [inline src/com/lilithsthrone/game/combat/Combat.java]: Allies could be a hashmap with a boolean value, true if ally.

Comment (CognitiveMist) [reply src/com/lilithsthrone/game/combat/Combat.java]: Testing for membership in a HashSet should be equally efficient.

**Comment by OhgunOrwell** [comment]
I bet it would be interesting if someone would have integrated the Press Turn System from the Megami Tensei series into this impressive rework.

**Pull request merged**